src/Symfony/Component/Serializer/Tests/SerializerTest.php

```diff
@@ -633,6 +638,77 @@ public function testDeserializeAndUnwrap()
+            ],
+            'object' => [
+                'foo' => 'The type of the "foo" attribute for class "Symfony\Component\Serializer\Tests\Fixtures\CollectErrorsObjectDummy" must be one of "int" ("string" given).',
+                'bar' => 'The type of the "bar" attribute for class "Symfony\Component\Serializer\Tests\Fixtures\CollectErrorsObjectDummy" must be one of "int" ("string" given).',
+                'baz' => 'Cannot create an instance of "Symfony\Component\Serializer\Tests\Fixtures\CollectErrorsObjectDummy" from serialized data because its constructor requires parameter "baz" to be present.',
+            ],
+            'variadicObject' => [
+                'args' => [
+                    [
+                        'foo' => 'The type of the "foo" attribute for class "Symfony\Component\Serializer\Tests\Fixtures\CollectErrorsVariadicObjectDummy" must be one of "int" ("string" given).',
+                    ],
+                    [
+                        'foo' => 'The type of the "foo" attribute for class "Symfony\Component\Serializer\Tests\Fixtures\CollectErrorsVariadicObjectDummy" must be one of "int" ("string" given).',
+                        'args' => 'Cannot create an instance of "Symfony\Component\Serializer\Tests\Fixtures\CollectErrorsVariadicObjectDummy" from serialized data because the variadic parameter "args" can only accept an array.',
+                    ],
+                ],
+            ],
+        ];
+
+        $extractor = new PropertyInfoExtractor([], [new PhpDocExtractor(), new ReflectionExtractor()]);
+        $serializer = new Serializer(
+            [
+                new DateTimeNormalizer(),
+                new ObjectNormalizer(null, null, null, $extractor),
+                new ArrayDenormalizer(),
+            ],
+            [
+                'json' => new JsonEncoder(),
+            ]
+        );
+
+        try {
+            $serializer->deserialize($jsonData, CollectErrorsDummy::class, 'json', [
+                Serializer::COLLECT_EXCEPTIONS => true,
+            ]);
+        } catch (AggregatedException $exception) {
+            $actual = ($f = static function (ExceptionInterface $exception) use (&$f) {
+                $actual = [];
+                foreach ($exception->getExceptions() as $data) {
+                    if ($data['exception'] instanceof AggregatedException) {
+                        $result = $f($data['exception']);
+                    } else {
+                        $result = $data['exception']->getMessage();
+                    }
+
+                    if (\array_key_exists($data['param'], $actual)) {
+                        $tmp = $actual[$data['param']];
+                        $actual[$data['param']] = [];
+                        $actual[$data['param']][] = $tmp;
+                        $actual[$data['param']][] = $result;
+                    } else {
+                        $actual[$data['param']] = $result;
+                    }
```

Comment:
I'm not a big fan of this, I wonder if the `param` should not be part of the `AggregatedException` instead.
I think you added this array with `param` and `exception` recently and it breaks the contract of the `AggregatedException::getExceptions()` method which is supposed to return an `ExceptionInterface[]`.

It might be a good idea to remove some complexity by keeping only one dimension the `AggregatedException`, its interface might looks like:
 * `AggregatedException::all(): array` the return type being `array<string, AggregableExceptionInterface>`
 * `AggregatedException::set(string $propertyName, AggregableExceptionInterface $exception): self`

So it could be used like:
```php
$exceptions = new AggregatedException();

foreach ($data as $attribute => $value) {
    try {
        $this->denormalize($value, $type, $format, $context);
    } catch (AggregableExceptionInterface $exception) {
        $exceptions->set($attribute, $exception);
    }
}
```

So that instead of dealing with an array like `['foo' => ['bar' => $exception, 'baz' => $otherException]]` it would be `['foo.bar' => $exception, 'foo.baz' => $otherException]`

What do you think ?

Reply:
I think it's a good idea. I did the implementation for that. Due to ambiguity of variadic arguments I used a Generator in `getExceptions` method.